File: 713-subarray-product-less-than-k/713-subarray-product-less-than-k.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int left = 0, prod = 1, count = 0;
        
        for(int right = 0; right < nums.length; right++){
            
            prod *= nums[right];
            
            while(prod >= k && left <= right){
                prod /= nums[left];
                left++;
            }
            count += right - left + 1;
        }
        
        return count;
    }
}

File: 713-subarray-product-less-than-k/713-subarray-product-less-than-k.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        start = 0
        ans = 0
        prod = 1
        
        if k <= 1:
            return 0
        
        for ind, val in enumerate(nums):
            prod *= val
            while prod >= k:
                prod //= nums[start]
                start += 1
            
            ans += ind - start + 1
        return ans